src/device-state.coffee

```diff
@@ -675,19 +676,25 @@ module.exports = class DeviceState extends EventEmitter
 		@applyUnblocker?()
 		return
 
-	triggerApplyTarget: ({ force = false, delay = 0, initial = false } = {}) =>
+	triggerApplyTarget: ({ force = false, delay = 0, initial = false, isFromApi = false } = {}) =>
 		if @applyInProgress
-			if !@scheduledApply?
+			if !@scheduledApply? || (isFromApi && @cancelDelay)
 				@scheduledApply = { force, delay }
+				if isFromApi
+					# Cancel promise delay if call came from api to prevent waiting due to backoff
+					@cancelDelay()
 			else
 				# If a delay has been set it's because we need to hold off before applying again,
 				# so we need to respect the maximum delay that has been passed
 				@scheduledApply.delay = Math.max(delay, @scheduledApply.delay)
 				@scheduledApply.force or= force
 			return
 		@applyInProgress = true
-		Promise.delay(delay)
+		new Promise (resolve, reject) =>
```

Comment:
I just realized we need to add a `.catch(->)` for this promise, otherwise there'll be an unhandled rejection when the promise is cancelled. There's a few ways to do this that I can think of, I would suggest something like:

```
@applyInProgress = true
@applyCancelled = false
new Promise (resolve, reject) =>
	setTimeout(resolve, delay)
	@cancelDelay = reject
.catch =>
	@applyCancelled = true
.then =>
	@cancelDelay = null
	if @applyCancelled
		log.info('Skipping applyTarget because of a cancellation')
		return
	log.info('Applying target state')
	@applyTarget({ force, initial })
.finally =>
...
```

Sorry I didn't spot this before...

Reply:
No worries, but you're right.
Why not place the catch below the then and save the @applyCanceled?

```
@applyInProgress = true
new Promise (resolve, reject) =>
	setTimeout(resolve, delay)
	@cancelDelay = reject
.then =>
	@cancelDelay = null
	log.info('Applying target state')
	@applyTarget({ force, initial })
.catch =>
        @cancelDelay = null
        log.info('Skipping applyTarget because of a cancellation')
.finally =>
...
```
This would come at the cost of setting cancelDelay null in two places in the code, but I think it might be better than introducing another variable.

Reply:
hmm but then on the other side my suggestion would also catch exceptions or errors from the then block lead to a log message that is not descriptive of what happend.

Reply:
Yup, that’s exactly what I was going to say :) after the then block we don’t know for sure where the error comes from. Alternatively you could use typed errors but it gets more complicated imo.

Reply:
I implemented your suggestion. Since we know if the promise is cancelled I simplified the test while doing that.